### PR TITLE
[Datumaro] Reducing nesting of tests

### PR DESCRIPTION
**Pull request by KochankovID**
### Motivation and context
Tests are complex.  It's necessary to reduce nesting. 
To achieve this goal was implemented Dataset.from_iterable(), labelCategories.from_iterable() 
and PointsCategories.from_iterable(). The format tests was simplified.

### How has this been tested?
It was tested using unittests.

### Checklist

- [x] I submit my changes into the `develop` branch
- [ ] I have added description of my changes into [CHANGELOG](https://github.com/opencv/cvat/blob/develop/CHANGELOG.md) file
- [ ] I have updated the [documentation](
  https://github.com/opencv/cvat/blob/develop/README.md#documentation) accordingly
- [x] I have added tests to cover my changes
- [ ] I have linked related issues ([read github docs](
  https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword))
- [ ] I have increased versions of npm packages if it is necessary ([cvat-canvas](https://github.com/opencv/cvat/tree/develop/cvat-canvas#versioning),
[cvat-core](https://github.com/opencv/cvat/tree/develop/cvat-core#versioning), [cvat-data](https://github.com/opencv/cvat/tree/develop/cvat-data#versioning) and [cvat-ui](https://github.com/opencv/cvat/tree/develop/cvat-ui#versioning))

### License

- [x] I submit _my code changes_ under the same [MIT License](
  https://github.com/opencv/cvat/blob/develop/LICENSE) that covers the project.
  Feel free to contact the maintainers if that's a concern.
- [ ] I have updated the license header for each file (see an example below)

```python
# Copyright (C) 2020 Intel Corporation
#
# SPDX-License-Identifier: MIT
```

### CODE REVIEW

File: datumaro/tests/test_coco_format.py [hunk 1]
```diff
@@ -40,17 +39,20 @@ def __iter__(self):
...
                         ]
                     ),
-                ])
-
-            def categories(self):
-                label_cat = LabelCategories()
-                label_cat.add('TEST')
-                return { AnnotationType.label: label_cat }
+                ]
+
```
Comment (nmanovic) [inline datumaro/tests/test_coco_format.py:43]: Codacy found an issue: [Trailing whitespace](https://app.codacy.com/manual/cvat/cvat/pullRequest?prid=5828054)

**Comment by nmanovic** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/cvat/cvat/pullRequest?prid=5828054&bid=19082074)

File: datumaro/datumaro/components/project.py [hunk 2]
```diff
@@ -315,6 +315,41 @@ def categories(self):
         return self._parent.categories()
 
 class Dataset(Extractor):
+    @classmethod
+    def from_iterable(cls, iterable, categories={}):
```
Comment (nmanovic) [inline datumaro/datumaro/components/project.py:319]: Codacy found an issue: [Dangerous default value {} as argument](https://app.codacy.com/manual/cvat/cvat/pullRequest?prid=5828054)

**Comment by nmanovic** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/cvat/cvat/pullRequest?prid=5828054&bid=19082074)

File: datumaro/tests/test_coco_format.py [hunk 3]
```diff
@@ -39,18 +38,20 @@ def __iter__(self):
...
                         ]
-                    ),
-                ])
-
-            def categories(self):
-                label_cat = LabelCategories()
-                label_cat.add('TEST')
-                return { AnnotationType.label: label_cat }
+                    )]
+
```
Comment (nmanovic) [inline datumaro/tests/test_coco_format.py:42]: Codacy found an issue: [Trailing whitespace](https://app.codacy.com/manual/cvat/cvat/pullRequest?prid=5828054)

**Comment by nmanovic** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/cvat/cvat/pullRequest?prid=5828054&bid=19082074)

File: datumaro/tests/test_coco_format.py [hunk 4]
```diff
@@ -39,18 +38,20 @@ def __iter__(self):
...
 
-            def categories(self):
-                label_cat = LabelCategories()
-                label_cat.add('TEST')
-                return { AnnotationType.label: label_cat }
+        # Creating categories
+        label_cat = LabelCategories()
+        label_cat.add('TEST')
+        categories = { AnnotationType.label : label_cat}
+
```
Comment (nmanovic) [inline datumaro/tests/test_coco_format.py:47]: Codacy found an issue: [Trailing whitespace](https://app.codacy.com/manual/cvat/cvat/pullRequest?prid=5828054)

File: datumaro/datumaro/components/project.py [hunk 5]
```diff
@@ -315,6 +315,46 @@ def categories(self):
...
+
+        Args:
+            iterable: Iterable object contains DatasetItems
+            categories (dict, optional): Dict of categories of datasets. Defaults to {}.
+
+        Returns:
+            Dataset: Dataset object
+        """
+
+        # Configuration of categories
```
Comment (zhiltsov-max) [inline datumaro/datumaro/components/project.py:330]: Please read about commenting your code https://www.elegantthemes.com/blog/wordpress/how-to-comment-your-code-like-a-pro-best-practices-and-good-habits

There is no need in commenting each line.

File: datumaro/datumaro/components/project.py [hunk 6]
```diff
@@ -315,6 +315,46 @@ def categories(self):
...
+        if not categories:
+            categories = {}
+
+        # Get Dataset with specified categories
+        dataset = Dataset(categories=categories)
+
+        #Get iterator
+        it = iter(iterable)
```
Comment (zhiltsov-max) [inline datumaro/datumaro/components/project.py:338]: This can be moved right into `for` below.

File: datumaro/datumaro/components/project.py [hunk 7]
```diff
@@ -315,6 +315,46 @@ def categories(self):
...
+
+        # Merging items
+        subsets = defaultdict(lambda: Subset(dataset))
+
+        for item in it:
+                existing_item = subsets[item.subset].items.get(item.id)
```
Comment (zhiltsov-max) [inline datumaro/datumaro/components/project.py:344]: I suggest to think about extracting common code. Actually, this function can be implemented with `from_extractors` by creating a simple `Extractor` subclass here.

File: datumaro/tests/test_coco_format.py [hunk 8]
```diff
@@ -39,18 +38,20 @@ def __iter__(self):
...
+        # Creating exemplary dataset
+        exemplary_dataset = Dataset.from_iterable(items_list, categories)
 
         dataset = Project.import_from(DUMMY_DATASET_DIR, 'coco') \
             .make_dataset()
 
-        compare_datasets(self, DstExtractor(), dataset)
+        compare_datasets(self, exemplary_dataset, dataset)
```
Comment (zhiltsov-max) [inline datumaro/tests/test_coco_format.py:54]: I suggest to stick to the naming used everywhere else with `source`, `target` or `actual` and `expected`.

File: datumaro/tests/test_coco_format.py [hunk 9]
```diff
@@ -39,18 +38,20 @@ def __iter__(self):
...
-                label_cat = LabelCategories()
-                label_cat.add('TEST')
-                return { AnnotationType.label: label_cat }
+        # Creating categories
+        label_cat = LabelCategories()
+        label_cat.add('TEST')
+        categories = { AnnotationType.label : label_cat}
+
+        # Creating exemplary dataset
+        exemplary_dataset = Dataset.from_iterable(items_list, categories)
```
Comment (zhiltsov-max) [inline datumaro/tests/test_coco_format.py:49]: Consider creating and passing the first argument in-place to further simplify code. There is no need in a separate variable for the list above.

File: datumaro/tests/test_coco_format.py (hunk 8 above)
Comment (zhiltsov-max) [reply datumaro/tests/test_coco_format.py:54]: In general, tests follow AAA-pattern https://medium.com/@pjbgf/title-testing-code-ocd-and-the-aaa-pattern-df453975ab80

File: datumaro/tests/test_coco_format.py [hunk 10]
```diff
@@ -39,18 +38,20 @@ def __iter__(self):
                                 ), label=0,
                                 id=2, group=2, attributes={'is_crowd': True}),
                         ]
-                    ),
-                ])
+                    )]
```
Comment (zhiltsov-max) [inline datumaro/tests/test_coco_format.py:41]: Please, keep formatting the same as everywhere else.

File: datumaro/tests/test_coco_format.py [hunk 11]
```diff
@@ -25,9 +25,11 @@
 
 class CocoImporterTest(TestCase):
     def test_can_import(self):
-        class DstExtractor(Extractor):
-            def __iter__(self):
-                return iter([
+        label_cat = LabelCategories()
+        label_cat.add('TEST')
+
+        expected_dataset = Dataset.from_iterable(
```
Comment (zhiltsov-max) [inline datumaro/tests/test_coco_format.py:31]: I'd better wrote this the following way:

```
expected = Dataset.from_iterable([
    DatasetItem(...),
    ...
], categories = { ... })
```

**Comment by nmanovic** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- datumaro/tests/test_coco_format.py  3
- datumaro/tests/test_mot_format.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/cvat/cvat/pullRequest?prid=5828054&bid=19082074)

File: datumaro/tests/test_transforms.py [hunk 12]
```diff
@@ -14,29 +14,22 @@
 
 class TransformsTest(TestCase):
     def test_reindex(self):
```
Comment (KochankovID) [inline datumaro/tests/test_transforms.py:16]: On my machine this test works. But I have another test in this section, that has stopped to work when I have rewritten it: "test_remap_labels"

**Comment by nmanovic** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- datumaro/tests/test_mot_format.py  1
- datumaro/tests/test_coco_format.py  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/cvat/cvat/pullRequest?prid=5828054&bid=19082074)

File: datumaro/datumaro/components/extractor.py [hunk 13]
```diff
@@ -72,6 +72,20 @@ def __eq__(self, other):
 class LabelCategories(Categories):
     Category = namedtuple('Category', ['name', 'parent', 'attributes'])
 
+    @classmethod
+    def from_iterable(cls, iterable):
```
Comment (zhiltsov-max) [inline datumaro/datumaro/components/extractor.py:76]: I can't see this method really simplifies what's already in the class constructor. How about providing an option to pass just a list labels in `Dataset.from_iterable()`?

File: datumaro/tests/test_cvat_format.py [hunk 14]
```diff
@@ -27,9 +27,7 @@ def test_can_detect_video(self):
         self.assertTrue(CvatImporter.detect(DUMMY_VIDEO_DATASET_DIR))
 
     def test_can_load_image(self):
-        class DstExtractor(Extractor):
-            def __iter__(self):
-                return iter([
+        expected_dataset = Dataset.from_iterable([
```
Comment (zhiltsov-max) [inline datumaro/tests/test_cvat_format.py:30]: The idea of the patch is reducing nesting. Can you adjust the indentation so it was actually true?

**Comment by zhiltsov-max** [comment]
Could you fill the PR description?

File: datumaro/datumaro/components/extractor.py (hunk 13 above)
Comment (KochankovID) [reply datumaro/datumaro/components/extractor.py:76]: There is such opportunity. You can pass a list labels' names or a list with labels' params. I thing it's more flexible and allow you to simplify more tests.

File: datumaro/tests/test_cvat_format.py (hunk 14 above)
Comment (KochankovID) [reply datumaro/tests/test_cvat_format.py:30]: I didn't catch what you mean here. Could you please describe in more details?

Comment (zhiltsov-max) [reply datumaro/tests/test_cvat_format.py:30]: I wrote it earlier. I'd like to see it like this:
```
expected = Dataset.from_iterable([
    DatasetItem(...),
    ...
], categories=['label_1', 'label_2', ...])
```

Currently, there are 2 extra levels of indentation for no reason.

Comment (KochankovID) [reply datumaro/tests/test_cvat_format.py:30]: I see only two differences between my code and your:
1) another name of variable. do you mean _ by saying "indentation" ? Do I need to replace it?
2) in categories. Lately, you have said to me to cope with categories by adding from_iterable to LableCategories. That is what I have done. 
Now I see, that you want me to rewrite Dataset.from_iterable. But how can I specify type of annotations and categories, if you are going just to pass a single list on categories names?
There are several types of categories, not only lableCategories.

Comment (KochankovID) [reply datumaro/tests/test_cvat_format.py:30]: And what you wrote earlier is different. You wrote:
expected = Dataset.from_iterable([
    DatasetItem(...),
    ...
], categories={ ... })

and here, for categories dict is used. 

That's why I didn't understand you.

Comment (zhiltsov-max) [reply datumaro/tests/test_cvat_format.py:30]: 1. By "indentation" I mean the number of leading spaces in lines of code, this is an established term. Currently, there is no changes in the PR regarding this, they may have failed to be pushed by git, it happens.

2. You're right about several types, but in practice we mostly need only labels, and this case is simple to distinguish by the type of argument or by another argument. You can come up with a better idea, if you have one. I think, it is the easiest way of declaring a dataset.

Comment (KochankovID) [reply datumaro/tests/test_cvat_format.py:30]: 1. Я не понимаю как ты хочешь сократить кол-во начальных пробелов в строке? В питоне они используются для определения программных блоков. 
Про вложенность: я могу вообще ее убрать, создавая списки не "на месте" а записывая в переменные, что я и хотел сделать, но ты сказал что лучше написать "на месте".
2. Я повторюсь, но скажу, что я просто сделал то, что сказал мне ты. Окей, я перепишу Dataset.from_iterable(). И перепишу тесты встоответсвии. Но вопервых, для некоторых тестов это будет сделать невозможно (там где используются не LableCategorries. А во вторых я все еще не понимаю что ты от меня хочешь, говоря "убрать вложенность" или кол-во пробелов в начале строки. Было бы здорово если бы ты взял пример моего теста полностью и переписал так, как ты хочешь это видеть и указал на изменения.

**Comment by nmanovic** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- datumaro/tests/test_coco_format.py  3
- datumaro/tests/test_mot_format.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/cvat/cvat/pullRequest?prid=5828054&bid=19082074)

File: datumaro/tests/test_transforms.py [hunk 15]
```diff
@@ -373,26 +338,25 @@ def __iter__(self):
...
 
         with self.assertRaises(Exception):
-            transforms.RandomSplit(SrcExtractor(), splits=[
+            transforms.RandomSplit(source_dataset, splits=[
                 ('train', -0.5),
                 ('test', 1.5),
             ])
 
     def test_remap_labels(self):
+        #! why this test brokes when I rewrite it on Dataset ?
```
Comment (zhiltsov-max) [inline datumaro/tests/test_transforms.py:359]: Please, clean the comment. This probably can be fixed with non-inplace implementation of the transform. Another way to fix is probably to make a dataset from the transform result.

**Comment by nmanovic** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- datumaro/tests/test_coco_format.py  3
- datumaro/tests/test_mot_format.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/cvat/cvat/pullRequest?prid=5828054&bid=19082074)

File: datumaro/tests/test_labelme_format.py [hunk 16]
```diff
@@ -29,101 +29,84 @@ def _test_save_and_load(self, source_dataset, converter, test_dir,
...
+            ),
+        ], categories={
+            AnnotationType.label: LabelCategories.from_iterable([
+                'label_2', 'label_3']),
+        })
 
         with TestDir() as test_dir:
-            self._test_save_and_load(SrcExtractor(),
+            self._test_save_and_load(
+                source_dataset,
```
Comment (nmanovic) [inline datumaro/tests/test_labelme_format.py:95]: Codacy found an issue: [Trailing whitespace](https://app.codacy.com/manual/cvat/cvat/pullRequest?prid=5828054)

**Comment by nmanovic** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- datumaro/tests/test_coco_format.py  3
- datumaro/tests/test_mot_format.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/cvat/cvat/pullRequest?prid=5828054&bid=19082074)

**Comment by nmanovic** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- datumaro/tests/test_coco_format.py  3
- datumaro/tests/test_mot_format.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/cvat/cvat/pullRequest?prid=5828054&bid=19082074)

**Comment by nmanovic** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- datumaro/tests/test_mot_format.py  1
- datumaro/tests/test_coco_format.py  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/cvat/cvat/pullRequest?prid=5828054&bid=19082074)

**Pull request merged**